lisp/ess-utils.el

```diff
@@ -446,13 +446,21 @@ etc.
 
 ;;;*;;; Font Lock
 
-(defun ess--extract-default-fl-keywords (keywords)
+(defun ess--extract-default-fl-keywords (keywords-alist)
   "Extract the t-keywords from `ess-font-lock-keywords'."
-  (delq nil (mapcar (lambda (c)
-                      (when (cdr c) (symbol-value (car c))))
-                    (if (symbolp keywords)
-                        (symbol-value keywords)
-                      keywords))))
+  (let ((keywords-alist (if (symbolp keywords-alist)
+                            (symbol-value keywords-alist)
+                          keywords-alist)))
+    ;; `keywords' can be either a font-lock keyword list starting with
+    ;; a regexp string or a form that should evaluate to such a
+    ;; keyword list
+    (delq nil (mapcar (lambda (cell)
+                        (when (cdr cell)
+                          (let ((keywords (symbol-value (car cell))))
+                            (if (stringp (car keywords))
```

Comment:
This won't work if keywords if keywords is a regexp or a function or a cell with car a function. See the Info 23.6.2 Search-based Fontification. 

I also wonder if the built-in font-lock functionality could be used here:

```
‘(eval . FORM)’
     Here FORM is an expression to be evaluated the first time this
     value of ‘font-lock-keywords’ is used in a buffer.  Its value
     should have one of the forms described in this table.
```

Reply:
Of course, good catch. The `eval` form seems relevant here, thanks.